components/net_traits/image_cache.rs

```diff
@@ -224,272 +217,203 @@ impl CompletedLoad {
-                        let img_metadata = ImageMetadata { width: dimensions.width,
-                                                           height: dimensions.height };
-                        for listener in &pending_load.listeners {
-                            listener.respond(ImageResponse::MetadataLoaded(img_metadata.clone()));
-                        }
-                        pending_load.metadata = Some(img_metadata);
-                    }
-                }
-            }
-            (FetchResponseMsg::ProcessResponseEOF(result), key) => {
-                debug!("received EOF for {:?}", key);
-                match result {
-                    Ok(()) => {
-                        let pending_load = self.pending_loads.get_by_key_mut(&msg.key).unwrap();
-                        pending_load.result = Some(result);
-                        let bytes = pending_load.bytes.mark_complete();
-                        let sender = self.decoder_sender.clone();
-                        debug!("async decoding {} ({:?})", pending_load.url, key);
+#[derive(Copy, Clone, PartialEq, Hash, Eq, Deserialize, Serialize)]
+pub enum UsePlaceholder {
+    No,
+    Yes,
+}
 
-                        self.thread_pool.execute(move || {
-                            let msg = decode_bytes_sync(key, &*bytes);
-                            sender.send(msg).unwrap();
-                        });
-                    }
-                    Err(_) => {
-                        debug!("processing error for {:?}", key);
-                        match self.placeholder_image.clone() {
-                            Some(placeholder_image) => {
-                                self.complete_load(msg.key, LoadResult::PlaceholderLoaded(
-                                        placeholder_image))
-                            }
-                            None => self.complete_load(msg.key, LoadResult::None),
-                        }
-                    }
-                }
-            }
-        }
-    }
+// ======================================================================
+// ImageCache implementation.
+// ======================================================================
 
-    // Handle a message from one of the decoder worker threads
-    fn handle_decoder(&mut self, msg: DecoderMsg) {
-        let image = match msg.image {
-            None => LoadResult::None,
-            Some(image) => LoadResult::Loaded(image),
-        };
-        self.complete_load(msg.key, image);
-    }
+pub struct ImageCache {
+    // Images that are loading over network, or decoding.
+    pending_loads: Mutex<AllPendingLoads>,
+
+    // Images that have finished loading (successful or not)
+    completed_loads: Mutex<HashMap<ServoUrl, CompletedLoad>>,
+
+    // The placeholder image used when an image fails to load
+    placeholder_image: Option<Arc<Image>>,
 
+    // Webrender API instance.
+    webrender_api: Mutex<webrender_traits::RenderApi>,
```

Comment:
Ok, thanks for the explanation.

I am a bit blocked trying to address this comment. I think I understand the issue, but I am not sure how to solve it.

So the `ImageCacheImpl` struct has a `webrender_traits::RenderApi` member and `webrender_traits::RenderApi` contains a `Cell<webrender_traits::ResourceId>`. The thing is that `Cell` implements `Send` but not `Sync` and I was solving this by wrapping the `RenderApi` member in its own Mutex. But now, removing this Mutex and wrapping the entire `ImageCacheImpl` in a Mutex makes the compiler unhappy:

```
rror[E0277]: the trait bound `std::cell::Cell<webrender_traits::ResourceId>: std::marker::Sync` is not satisfied in `image_cache::ImageCacheImpl`
   --> /Volumes/mozilladev/servo/components/net/image_cache.rs:406:6
    |
406 | impl ImageCache for ImageCacheImpl {
    |      ^^^^^^^^^^ within `image_cache::ImageCacheImpl`, the trait `std::marker::Sync` is not implemented for `std::cell::Cell<webrender_traits::ResourceId>`
    |
    = note: `std::cell::Cell<webrender_traits::ResourceId>` cannot be shared between threads safely
    = note: required because it appears within the type `webrender_traits::RenderApi`
    = note: required because it appears within the type `image_cache::ImageCacheImpl`
    = note: required by `net_traits::image_cache::ImageCache`

error[E0277]: the trait bound `std::cell::Cell<()>: std::marker::Sync` is not satisfied in `image_cache::ImageCacheImpl`
   --> /Volumes/mozilladev/servo/components/net/image_cache.rs:406:6
    |
406 | impl ImageCache for ImageCacheImpl {
    |      ^^^^^^^^^^ within `image_cache::ImageCacheImpl`, the trait `std::marker::Sync` is not implemented for `std::cell::Cell<()>`
    |
    = note: `std::cell::Cell<()>` cannot be shared between threads safely
    = note: required because it appears within the type `std::marker::PhantomData<std::cell::Cell<()>>`
    = note: required because it appears within the type `ipc_channel::platform::OsIpcSender`
    = note: required because it appears within the type `ipc_channel::ipc::IpcSender<webrender_traits::ApiMsg>`
    = note: required because it appears within the type `webrender_traits::RenderApi`
    = note: required because it appears within the type `image_cache::ImageCacheImpl`
    = note: required by `net_traits::image_cache::ImageCache`
```

@jdm should I keep the `Mutex` for the `webrender_traits::RenderApi` member? Or is there a better way to solve this?